### PR TITLE
Harden MCP spec coverage and release validation

### PR DESCRIPTION
**Pull request by leehack**
## Summary

- Harden MCP 2026-07-28 request-scoped cancellation, OAuth ownership, reconnect, and terminal SSE cleanup without changing MCP 2025-11-25 cancellation semantics.
- Upgrade built-in JSON Schema validation to Draft 2020-12 with declared Draft 7 compatibility, same-document reference handling, and an explicit no-network policy.
- Validate MCP sampling tool-use/result sequences, tool input/output schemas, and official conformance scenario inventories.
- Add pinned official JSON Schema Test Suite gates, lifecycle stress CI, expanded TypeScript/Python interoperability, and real Chrome/Flutter Web coverage.
- Refresh examples, release runbooks, coverage matrices, changelogs, and package-facing documentation to use full MCP specification names.

## Why

High line coverage and green conformance runs were not catching several real-use lifecycle and interoperability failures. In particular, cancellation during OAuth/HTTP setup, terminal SSE responses with open bodies, repeated browser requests, sampling tool-result ordering, and schema-dialect behavior needed use-case and edge-case tests tied directly to the specification.

This change turns those cases into durable regression, browser, interop, and pinned upstream-suite gates.

## Compatibility

There are **no public API removals, renames, or changes to existing required call parameters relative to `main`**. Protocol constant values and the Dart SDK floor (`^3.5.0`) are unchanged.

Normal constructor-based SDK callers and custom transports that implement only `Transport` remain source-compatible. There is one narrow Dart interface-compatibility caveat: downstream types that explicitly `implements StartSseOptions` or `implements StreamableHttpClientTransport` must add the new cancellation members, because Dart includes public class members in the implicit interface. The additions are otherwise optional: `RequestCancellationAwareTransport` and the new `StartSseOptions` cancellation hooks. The public `JsonObject.extra` API remains preserved.

There are material, intentional behavior-tightening changes:

- `CreateMessageRequest.fromJson()` and `toJson()` now reject MCP-invalid tool-use/result roles, ordering, mixed content, duplicate IDs, and missing or mismatched results.
- Built-in validation now uses JSON Schema Draft 2020-12 by default and declared Draft 7 semantics. Validation outcomes and error text/path may differ from the old incomplete validator. Unsupported dialects/custom vocabularies and unresolved external references are rejected without network access; evaluation is bounded.
- `JsonSchema.fromJson()` may return `JsonAny` instead of a primitive or array wrapper when a typed wrapper would discard valid keywords or tuple `items`. Code casting parsed schemas by runtime subtype should use the wire schema/API instead.
- MCP 2026-07-28 Streamable HTTP cancellation closes only the matching POST response stream when supported instead of sending `notifications/cancelled`. MCP 2025-11-25 and transports without the optional capability retain notification cancellation.
- Duplicate active stateless request IDs now throw `StateError`; terminal SSE responses settle immediately and ignore later frames; transport close interrupts pending OAuth/request lifecycle work.
- The runtime dependency set adds `json_schema ^5.2.2` and its transitive footprint.

Migration, if affected: fix invalid sampling histories or schemas, avoid matching exact validator messages or assuming parsed schema runtime subtypes, and treat transport close as cancellation of pending lifecycle operations. Normal compliant flows remain compatible.

## Validation

- `dart format --output=none --set-exit-if-changed .`
- `dart analyze`
- Root tests: **1,602 passed**
- Official MCP example audit: **128/128 parsed**
- Official MCP document inventory: **31/31**, no missing/stale/invalid entries
- JSON Schema Test Suite: **1,242 Draft 2020-12** and **904 Draft 7** supported assertions, zero failures
- MCP 2025-11-25 client conformance: **221/221**; server scenarios all green
- MCP 2026-07-28 server conformance: **40/40**; client scenarios all green
- Bidirectional TypeScript and Python SDK interoperability
- Chrome transport cancellation/recovery and repeated MCP 2025-11-25/MCP 2026-07-28 requests
- Built Flutter Web UI: eight consecutive tool calls, reconnect, and post-reconnect call; all 13 MCP POSTs returned 200 with zero browser errors
- Flutter, Jaspr, Anthropic, Gemini, fetch, and CLI example/package gates
- `actionlint`
- `dart pub publish --dry-run` and CLI publish-tree dry-run
- `pana`: **160/160**

## Release follow-up

The final MCP 2026-07-28 tag is still the release boundary. On day 0, compare the pinned schemas, examples, conformance inventory, and normative prose against the final tag, then rerun the checked-in release gate.

### CODE REVIEW

File: doc/quick-reference.md [hunk 1]
```diff
@@ -275,7 +276,10 @@ final schema = JsonSchema.object(
 );
 ```
 
-The SDK preserves supported JSON Schema 2020-12 constructs. Validate business
+The SDK validates JSON Schema Draft 2020-12 by default and accepts an explicitly
+declared Draft 7 schema for MCP 2025-11-25 compatibility. Same-document fragment
+references are supported; relative and network references, `$dynamicRef`,
+unsupported dialects, and custom vocabularies are rejected. Validate business
 rules in the callback as well as describing inputs in the schema.
```
[the remark this answers is not in the document]
Comment (leehack) [reply doc/quick-reference.md:283]: Fixed in c9bb3e8. The quick reference now distinguishes supported same-document `$ref` and `$dynamicRef` resolution from rejected relative/network references, unsupported dialects, and custom vocabularies.

File: lib/src/shared/protocol.dart [hunk 2]
```diff
@@ -2072,6 +2073,25 @@ abstract class Protocol {
...
+            _onerror(
+              StateError(
+                "Failed to cancel transport request $messageId: $e",
+              ),
+            );
+          });
+        }
+        completer.completeError(errorReason);
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (leehack) [reply lib/src/shared/protocol.dart:2097]: Kept the early return intentionally and clarified it in c9bb3e8. MCP 2026-07-28 defines Streamable HTTP cancellation by closing the matching POST response stream, whereas `notifications/cancelled` is the stdio binding; sending the legacy notification as a fallback would mix protocol-era semantics (https://modelcontextprotocol.io/specification/draft/basic/utilities/cancellation and https://modelcontextprotocol.io/specification/draft/basic/transports). The built-in HTTP transport registers request cancellation synchronously before its first await, so `false` means the request is no longer active/cancellable rather than a registration race. The new regression test proves that path never emits the removed legacy notification.

File: lib/src/server/streamable_https.dart [hunk 3]
```diff
@@ -1191,29 +1191,50 @@ class StreamableHTTPServerTransport
       headers.forEach((key, value) {
         res.headers.set(key, value);
       });
```
[the remark this answers is not in the document]
Comment (leehack) [reply lib/src/server/streamable_https.dart:1194]: No additional write is needed here: `handleRequest` sets `req.response.bufferOutput = false` for every incoming request before dispatching GET to `_handleGetRequest`, so the `HttpResponse` passed to `_replayEvents` is already unbuffered. The replay path also queues the complete replay batch before its first await, preserving delivery order with live events.

File: lib/src/types/sampling.dart [hunk 4]
```diff
@@ -106,6 +106,102 @@ List<SamplingContent> _asSamplingContentBlocks(
...
+
+    final missing = useIds.toSet().difference(resultIds.toSet());
+    final unexpected = resultIds.toSet().difference(useIds.toSet());
+    if (missing.isNotEmpty || unexpected.isNotEmpty) {
+      fail(
+        'tool results at message ${index + 1} must match the preceding tool '
+        'use IDs exactly (missing: ${missing.toList()}, unexpected: '
+        '${unexpected.toList()})',
+      );
+    }
```
[the remark this answers is not in the document]
Comment (leehack) [reply lib/src/types/sampling.dart]: Fixed in 31f5634. Missing and unexpected tool IDs are now converted to lists and sorted before formatting, and a regression test verifies stable ordering from deliberately reversed inputs.
